### PR TITLE
_transactions2, Part 4: Client Aware Managed Timestamp Services

### PR DESCRIPTION
**Pull request by jeremyk-91**
**Goals (and why)**:
- Have a version of the timestamp service interface that allows users to provide an ID; we'll do the right thing from then on (here, giving out timestamps with a stable residue modulo-16).

**Implementation Description (bullets)**:
- Implement `getFreshTimestampForClient`. This works by retrieving 16 timestamps from the timestamp service and then getting one that matches our residue according to a `NumericPartitionAllocator`.

**Testing (What was existing testing like?  What have you done to improve it?)**:
- This is new code, and it comes tested.

**Concerns (what feedback would you like?)**:
- Does it make sense to centralise the logic for deciding on how to advance the clock in timelock?
- I made a decision that having the memorised information be destroyed at the time of a leader election is reasonable; is this the case?
- We grab 16 timestamps at a shot and then find the one we want. There are ways to do this while being a bit less wasteful, but I thought this is fine as timestamps are longs.

**Where should we start reviewing?**: `ClientAwareManagedTimestampService`

**Priority (whenever / two weeks / yesterday)**: this week / early next

---
This change is [Reviewable](https://reviewable.io/reviews/palantir/atlasdb/3527)

### CODE REVIEW

**Comment by j-baker** [comment]
grabbing 16 timestamps at a shot - fine for timestamp, but note that it breaks some other assumptions - e.g. that the targeted sweep queues are dense? Is the plan for this written down anywhere?

File: timelock-impl/src/main/java/com/palantir/atlasdb/timelock/transaction/timestamp/DelegatingClientAwareManagedTimestampService.java [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+                    timestampRange,
+                    targetResidue,
+                    NUM_PARTITIONS);
+            if (relevantTimestamp.isPresent()) {
+                return relevantTimestamp.getAsLong();
+            }
+
+            // Not a bug - getFreshTimestamps is permitted to return less than the number of timestamps asked for,
+            // so this case is possible.
+            log.info("The timestamp range we received from the underlying timestamp service - {} -"
```
Comment (gmaretic) [inline timelock-impl/src/main/java/com/palantir/atlasdb/timelock/transaction/timestamp/DelegatingClientAwareManagedTimestampService.java]: Seems reasonable, even if getFreshTimestamps is consistently returning fewer then NUM_PARTITIONS timestamps, the probability of consistently not having matches falls very quickly with the number of iterations.

File: timelock-impl/src/test/java/com/palantir/atlasdb/timelock/transaction/timestamp/DelegatingClientAwareManagedTimestampServiceTest.java [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+    private DelegatingClientAwareManagedTimestampService service = new DelegatingClientAwareManagedTimestampService(
+            allocator, timestamps);
+
+    @After
+    public void verifyNoFurtherInteractions() {
+        verifyNoMoreInteractions(timestamps, allocator);
+    }
+
+    @Test
+    public void getsFreshTimestampsFromDelegate() {
```
[the remark this answers is not in the document]
Comment (gmaretic) [reply timelock-impl/src/test/java/com/palantir/atlasdb/timelock/transaction/timestamp/DelegatingClientAwareManagedTimestampServiceTest.java]: I guess someone could change the `delegate()` implementation for this to fail

File: timelock-impl/src/test/java/com/palantir/atlasdb/timelock/transaction/timestamp/DelegatingClientAwareManagedTimestampServiceTest.java [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+    @Test
+    public void makesRequestsAgainIfTimestampRangeDoesNotIncludeCorrectModuli() {
+        when(allocator.getRelevantModuli(UUID_ONE)).thenReturn(MODULUS_ONE);
+        when(allocator.getRelevantModuli(UUID_TWO)).thenReturn(MODULUS_TWO);
+        when(timestamps.getFreshTimestamps(anyInt()))
+                .thenReturn(TIMESTAMP_ZERO)
+                .thenReturn(TIMESTAMP_ZERO)
+                .thenReturn(TIMESTAMP_RANGE);
+
+        assertThat(service.getFreshTimestampForClient(UUID_ONE)).isEqualTo(MODULUS_ONE_TIMESTAMP_IN_RANGE);
```
Comment (gmaretic) [inline timelock-impl/src/test/java/com/palantir/atlasdb/timelock/transaction/timestamp/DelegatingClientAwareManagedTimestampServiceTest.java]: I'd use `UUID_TWO` here to get slightly better coverage

File: timelock-impl/src/main/java/com/palantir/atlasdb/timelock/transaction/client/NumericPartitionAllocator.java [hunk 4]
```diff
@@ -0,0 +1,28 @@
+/*
+ * Copyright 2018 Palantir Technologies, Inc. All rights reserved.
+ *
+ * Licensed under the BSD-3 License (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://opensource.org/licenses/BSD-3-Clause
```
Comment (felixdesouza) [inline timelock-impl/src/main/java/com/palantir/atlasdb/timelock/transaction/client/NumericPartitionAllocator.java]: - [ ] License

File: timelock-impl/src/main/java/com/palantir/atlasdb/timelock/transaction/timestamp/ClientAwareManagedTimestampService.java [hunk 5]
```diff
@@ -0,0 +1,40 @@
+/*
+ * Copyright 2018 Palantir Technologies, Inc. All rights reserved.
+ *
+ * Licensed under the BSD-3 License (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://opensource.org/licenses/BSD-3-Clause
```
Comment (felixdesouza) [inline timelock-impl/src/main/java/com/palantir/atlasdb/timelock/transaction/timestamp/ClientAwareManagedTimestampService.java]: - [ ] License

File: timelock-impl/src/main/java/com/palantir/atlasdb/timelock/transaction/timestamp/DelegatingClientAwareManagedTimestampService.java [hunk 6]
```diff
@@ -0,0 +1,78 @@
+/*
+ * Copyright 2018 Palantir Technologies, Inc. All rights reserved.
+ *
+ * Licensed under the BSD-3 License (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://opensource.org/licenses/BSD-3-Clause
```
Comment (felixdesouza) [inline timelock-impl/src/main/java/com/palantir/atlasdb/timelock/transaction/timestamp/DelegatingClientAwareManagedTimestampService.java]: - [ ] License

File: timelock-impl/src/test/java/com/palantir/atlasdb/timelock/transaction/timestamp/DelegatingClientAwareManagedTimestampServiceTest.java [hunk 7]
```diff
@@ -0,0 +1,115 @@
+/*
+ * Copyright 2018 Palantir Technologies, Inc. All rights reserved.
+ *
+ * Licensed under the BSD-3 License (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://opensource.org/licenses/BSD-3-Clause
```
Comment (felixdesouza) [inline timelock-impl/src/test/java/com/palantir/atlasdb/timelock/transaction/timestamp/DelegatingClientAwareManagedTimestampServiceTest.java]: - [ ] License

File: timelock-impl/src/test/java/com/palantir/atlasdb/timelock/transaction/timestamp/DelegatingClientAwareManagedTimestampServiceTest.java [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+    private static final UUID UUID_TWO = UUID.randomUUID();
+
+    private static final TimestampRange TIMESTAMP_RANGE = TimestampRange.createInclusiveRange(
+            DelegatingClientAwareManagedTimestampService.NUM_PARTITIONS,
+            DelegatingClientAwareManagedTimestampService.NUM_PARTITIONS + 2);
+    private static final long MODULUS_ONE_TIMESTAMP_IN_RANGE
+            = DelegatingClientAwareManagedTimestampService.NUM_PARTITIONS + 1;
+    private static final TimestampRange TIMESTAMP_ZERO = TimestampRange.createInclusiveRange(0, 0);
+
+    private NumericPartitionAllocator<UUID> allocator = mock(NumericPartitionAllocator.class);
```
[the remark this answers is not in the document]
Comment (jeremyk-91) [reply timelock-impl/src/test/java/com/palantir/atlasdb/timelock/transaction/timestamp/DelegatingClientAwareManagedTimestampServiceTest.java]: Yep, these can be final. I was probably thinking of initialising these in a setUp() for some reason but turned out not to need it.

**Comment by jeremyk-91** [comment]
@j-baker The plan to take 16 at a time was mentioned in the internal RFC for this project, as found on internal remarkable documentation platform.

I don't think we have documented the consequences (e.g. density of rows in Targeted Sweep, update frequencies for Schema Leases and for the transactions2 migration) anywhere though

**Pull request merged**